src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs

```diff
@@ -99,13 +95,16 @@ public async Task AuthenticateAsync(IConnection connection, ConnectionDescriptio
         {
             Ensure.IsNotNull(connection, nameof(connection));
             Ensure.IsNotNull(description, nameof(description));
-
             using (var conversation = new SaslConversation(description.ConnectionId))
             {
-                var currentStep = _mechanism.Initialize(connection, conversation, description);
-
+                var currentStep = _mechanism.Initialize( connection, conversation, description);
                 var command = CreateStartCommand(currentStep);
-                while (true)
+                var speculativeAuthenticateResult = description.IsMasterResult.SpeculativeAuthenticate;
+                if (speculativeAuthenticateResult != null)
+                {
+                    currentStep = Transition(conversation, currentStep, speculativeAuthenticateResult, out command);
+                }
```

Comment:
Add a blank line here to visually separate "consuming the spec auth result" from the loop.